src/plugins/integrations/HttpIntegration.ts

```diff
@@ -48,73 +51,77 @@ class HttpIntegration implements Integration {
+          const queryParams = path.split('?').length > 1 ? path.split('?')[1] : '';
+
+          if (!HttpIntegration.isValidUrl(host)) {
+            return request.apply(this, [options, callback]);
+          }
 
-        const tracer = ThundraTracer.getInstance();
-        const method = (options.method || 'GET').toUpperCase();
-        options = typeof options === 'string' ? url.parse(options) : options;
-        const host = options.hostname || options.host || 'localhost';
-        const path = options.path || options.pathname || '/';
-        const queryParams = path.split('?').length > 1 ? path.split('?')[1] : '';
+          const parentSpan = tracer.getActiveSpan();
+          const span = tracer._startSpan(host + path, {
+            childOf: parentSpan,
+            domainName: DomainNames.API,
+            className: ClassNames.HTTP,
+            disableActiveStart: true,
+          });
+
+          if (!(Utils.getConfiguration(envVariableKeys.DISABLE_SPAN_CONTEXT_INJECTION) === 'true')) {
+            const headers = options.headers ? options.headers : {};
+            tracer.inject(span.spanContext, opentracing.FORMAT_TEXT_MAP, headers);
+            options.headers = headers;
+          }
 
-        if (!HttpIntegration.isValidUrl(host)) {
+          span.addTags({
+            [SpanTags.OPERATION_TYPE]: 'CALL',
+            [SpanTags.SPAN_TYPE]: SpanTypes.HTTP,
+            [HttpTags.HTTP_METHOD]: method,
+            [HttpTags.HTTP_HOST]: host,
+            [HttpTags.HTTP_PATH]: path.split('?')[0],
+            [HttpTags.HTTP_URL]: host + path,
+            [HttpTags.QUERY_PARAMS]: queryParams,
+            [SpanTags.TOPOLOGY_VERTEX]: true,
+            [SpanTags.TRIGGER_DOMAIN_NAME]: LAMBDA_APPLICATION_DOMAIN_NAME,
+            [SpanTags.TRIGGER_CLASS_NAME]: LAMBDA_APPLICATION_CLASS_NAME,
+            [SpanTags.TRIGGER_OPERATION_NAMES]: [tracer.functionName],
+          });
+
+          const req = request.call(this, options, callback);
+
+          req.on('socket', () => {
+            if (req.listenerCount('response') === 1) {
+              req.on('response', (res: any) => res.resume());
+            }
+          });
+
+          req.on('response', (res: any) => {
+            span.setTag(HttpTags.HTTP_STATUS, res.statusCode);
+            res.on('end', () => span.close());
+            span.close();
+          });
+
+          req.on('error', (err: any) => {
+            const parseError = Utils.parseError(err);
+
+            span.setTag('error', true);
+            span.setTag('error.kind', parseError.errorType);
+            span.setTag('error.message', parseError.errorMessage);
+            span.setTag('error.stack', parseError.stack);
+            span.setTag('error.code', parseError.code);
+
+            span.close();
+          });
+
+          return req;
+        } catch (error) {
+          ThundraLogger.getInstance().error(error);
           return request.apply(this, [options, callback]);
```

Comment:
Can we be sure that in case of error `request.apply` is not called twice

Reply:
In case of error in the wrapped function error callback is called. I verified, it is not called twice.